### DISCUSSION TITLE
Info 로그 찍는 범위에 대한 질문

### DISCUSSION THREAD

**Discussion by kang20** [toduck-App/toduck-backend#24] [Q&A]
## 회의에서 정리된 내용

- **Info 로그**는 모든 메서드 호출 및 반환 시 각각 1번씩 기록하며, 호출 시에는 인자, 반환 시에는 return 값을 함께 로깅한다.

## 문제점

현재 방식으로는 한 번의 API 호출에 너무 많은 로그가 기록됩니다.  
평균적으로 하나의 API 호출에  평균 10개의 로그가 생성되며, 이는 비효율적이라고 생각됩니다.

## 해결 방안

이 문제를 해결하기 위해 몇 가지 방법을 고려해보았습니다:

### 1. 현 상태를 유지한다

- **장점:** 모든 메서드 호출에 대한 상세 정보를 확인할 수 있다.
- **단점:** 너무 많은 로그가 쌓여 가독성이 떨어지고, 분석이 어려워질 수 있다.

### 2. Controller에서만 로그를 남긴다

- **의미:** Controller에 남기는 로그는 지표적으로 의미가 있다고 판단됩니다.
    - **API 호출 수 추적:** 각 API의 호출 빈도를 기록할 수 있으며, 이는 성능 개선에 도움이 됩니다.
    - **API 수행 시간 추적:** Controller에서 호출부터 결과 반환까지의 시간을 측정해 성능을 분석할 수 있습니다.
    - **사용자 행동 추적:** Controller의 매개변수만으로도 충분히 사용자 행동을 파악할 수 있습니다.
- **장점:**
    - log 가 효율적으로 쌓이고 가독성이 높아짐
 - **단점:** 위험한 장애 상황에 신속한 대처가 힘들 수 있습니다.
 
### 3. Controller에서만 로그를 남기되, 도메인 계층 정보도 함께 기록한다

- **방법:** Controller에서만 로그를 남기지만, Service와 같은 도메인 계층에서 수행된 클래스 이름과 메서드 이름을 `ThreadLocal`에 저장하고, Controller 반환 시에 이 정보를 한 번에 로그로 출력하는 방법입니다.
- **장점:**
    - Service나 도메인 계층에서 심각한 에러가 발생했을 때 에러 추적이 용이합니다.
    - 전체 호출 과정을 한번에 파악할 수 있습니다.
- **단점:** Controller의 반환 로그가 다소 verbose해질 수 있습니다.

## 추가 설명

Custom 에러가 발생할 경우, 성공 log가 호출되지 않고 에러 발생 지점에서 클래스 이름과 메서드 이름이 출력되도록 현재 PR (#23)에서 구현해두었습니다.

**Comment by Seol-JY** [comment]
현재 방식은 controller, usecase, service, repository 레이어 모두에서 로그를 찍는 건가요?

**Reply by kang20** [reply]
넵 저희가 bean으로 등록한 모든 component에 찍습니다

**Reply by Seol-JY** [reply]
@kang20 3번 방식대로 하되, 컨트롤러정보와 유즈케이스 정보만 로그로 노출시키는건 어떤가요?

**Reply by kang20** [reply]
3번 방식은 컨트롤러 호출시 한번 Info 로그를 찍습니다 ( 매게인자와 같이)
그리고 Controller 메서드가 반환될 때 return 값이랑 Info 로그를 찍습니다 (여기까진 2번과 동일)
근데 3번 방식은 API 호출 될 때 UseCase 나 Service Repository 가 호출 되었던 정보를 return 될 때 한번에 보여주는 방식으로 생각을 했습니다.

@Seol-JY  컨틀로러 정보와 유즈케이스 정보만 로그로 노출한다는 말씀이 매게 인자도 포함인가요 아님 클래스 명과 메서드명만을 포함하는건가요?

**Reply by Junad-Park** [reply]
@kang20 현 상태에서는 확실히 너무 많은 로그가 쌓여서 분석이 어려울 것 같다는 점에 동의 합니다! 저도 3번 방식이 에러 추적등을 고려했을 때 적절한 것 같습니다!

**Reply by Seol-JY** [reply]
> 3번 방식은 컨트롤러 호출시 한번 Info 로그를 찍습니다 ( 매게인자와 같이) 그리고 Controller 메서드가 반환될 때 return 값이랑 Info 로그를 찍습니다 (여기까진 2번과 동일) 근데 3번 방식은 API 호출 될 때 UseCase 나 Service Repository 가 호출 되었던 정보를 return 될 때 한번에 보여주는 방식으로 생각을 했습니다.
> 
> @Seol-JY 컨틀로러 정보와 유즈케이스 정보만 로그로 노출한다는 말씀이 매게 인자도 포함인가요 아님 클래스 명과 메서드명만을 포함하는건가요?

@kang20 매게인자와 반환값 모두 말하는 건 맞고, service와 reposiotry에 대한 정보를 담을 필요가 없다는 뜻이었습니다!
